# 내 프로젝트 v1 - 설비의 운전 조건 

> **v1 = 첫 번째 판.** 완성본이 아니라 오늘 만드는 첫 판이라는 뜻입니다.
>
> ⭐ **실습 16 · 17 · 18 · 19를 한 노트북에 담은 견본입니다.** 오늘 네 칸이 전부 여기 들어 있어요.
>
> **이것은 견본입니다.** 화면으로 함께 만든 것을 그대로 담았습니다.
> 여러분은 **본인 데이터로** 같은 순서를 한 벌 만드시면 됩니다. 이 파일은 고치지 말고 열어보기만 하세요.

| 실습 | 이 노트북의 어디 |
|---|---|
| 실습 16 - 문제 정의서 | 문제 정의 · 쓰는 데이터 · 한계 |
| 실습 17 - 모델링 계획표 | 모델링 계획 |
| 실습 18 - 모델링 v1 | 문법 노트 · 내 갈래 · 코드 · 받은 코드 확인 |
| 실습 19 - 결과 한 장 | v1 결과 · 다음에 할 일 |

> 실습마다 맨 끝에 **도전** 칸이 하나씩 붙어 있습니다. 도전은 안 하셔도 되지만, **네 개 다 있으면 포트폴리오에 그대로 실립니다.**


# 내 프로젝트 v1

# 실습 16 - 문제 정의서

## 문제 정의

**한 문장** : 설비의 운전 조건(온도·회전수·토크·공구마모)으로, 실제 고장이 나기 전에 고장 가능성이 높은 상태를 미리 골라낸다.

| 물음 | 내 답 |
|---|---|
| 누가 쓰나 | 설비 유지보수(정비) 담당자 |
| 무엇을 결정하나 | 이 설비를 지금 멈춰서 점검·정비할지, 그대로 가동할지 |
| 언제 결정하나 | 센서값(온도·회전수·토크·공구마모)이 들어오는 가동 중, 실제 고장이 나기 전 |
| 이 판단이 늦으면 | 가동 중 갑자기 멈춰서(다운타임) 생산이 끊기고, 부품·공구 손상이 커지며 수리 비용과 정지 시간이 함께 늘어난다 |

### 면접용 30초 설명

저는 설비가 실제로 고장 나기 전에, 지금 들어오는 온도나 회전수, 토크(회전할 때 걸리는 힘), 공구마모 같은 운전 데이터를 보고 고장 가능성이 높은 상태를 미리 찾아내는 문제를 다뤘습니다. 이걸 쓰는 사람은 설비 유지보수, 그러니까 정비를 담당하는 담당자이고, 이 예측으로 설비를 지금 멈춰서 점검할지 아니면 그대로 가동할지를 결정하게 됩니다. 이 판단은 설비가 돌아가는 동안, 센서값이 들어오는 그 시점에 실제 고장이 나기 전에 미리 내려져야 의미가 있습니다. 만약 이 판단이 늦어지면 설비가 갑자기 멈추는 다운타임, 즉 생산이 끊기는 정지 시간이 생기고, 부품이나 공구 손상까지 커져서 수리 비용과 멈춰 있는 시간이 함께 늘어납니다.

## Step 1. 데이터 부르고 알아보기

In [16]:
import pandas as pd

df = pd.read_csv("../../data/16_machine-failure.csv")

print("행 수, 열 수:", df.shape)
df.head(10)

행 수, 열 수: (10000, 7)


,Type,Air temperature [K],Process temperature [K],Rotational speed [rpm],Torque [Nm],Tool wear [min],Machine failure
0,M,298.1,308.6,1551,42.8,0,정상
1,L,298.2,308.7,1408,46.3,3,정상
2,L,298.1,308.5,1498,49.4,5,정상
3,L,298.2,308.6,1433,39.5,7,정상
4,L,298.2,308.7,1408,40.0,9,정상
5,M,298.1,308.6,1425,41.9,11,정상
6,L,298.1,308.6,1558,42.4,14,정상
7,L,298.1,308.6,1527,40.2,16,정상
8,M,298.3,308.7,1667,28.6,18,정상
9,M,298.5,309.0,1741,28.0,21,정상


### 확인한 것

- `Machine failure`가 결과 열이다 (정상 9661건 / 고장 339건, 3.39%)
- `Type`(L/M/H)만 범주형이고, 나머지 5개(`Air temperature [K]`, `Process temperature [K]`, `Rotational speed [rpm]`, `Torque [Nm]`, `Tool wear [min]`)는 조건이 되는 숫자 열이다

## Step 2. 쓰는 데이터

| 항목 | 내용 |
|---|---|
| 출처 | [수업에서 받은 정제본] |
| 규모 | [10000행 x 7열] |
| 입력으로 쓸 열 | [숫자로 된 측정 열 5개] |
| 결과 열 | [Machine failure] |
| 결과 열의 값 | [정상 / 고장] |
| 기준선이 될 숫자 | [고장 339건 (3.39%)] |

## Step 3. 한계

In [17]:
num_cols = ["Air temperature [K]", "Process temperature [K]", "Rotational speed [rpm]", "Torque [Nm]", "Tool wear [min]"]

print("결측치 총합:", df.isna().sum().sum())
print("중복 행 수:", df.duplicated().sum())
print()

print("숫자 열 간 상관관계 (일부):")
print(df[num_cols].corr().round(3))
print()

시험용_고장건수 = int(round(339 * 0.2))
print("시험용 20%로 나눴을 때 남는 고장 건수(대략):", 시험용_고장건수, "건")

결측치 총합: 0
중복 행 수: 0

숫자 열 간 상관관계 (일부):
                         Air temperature [K]  Process temperature [K]  \
Air temperature [K]                    1.000                    0.876   
Process temperature [K]                0.876                    1.000   
Rotational speed [rpm]                 0.023                    0.019   
Torque [Nm]                           -0.014                   -0.014   
Tool wear [min]                        0.014                    0.013   

                         Rotational speed [rpm]  Torque [Nm]  Tool wear [min]  
Air temperature [K]                       0.023       -0.014            0.014  
Process temperature [K]                   0.019       -0.014            0.013  
Rotational speed [rpm]                    1.000       -0.875            0.000  
Torque [Nm]                              -0.875        1.000           -0.003  
Tool wear [min]                           0.000       -0.003            1.000  

시험용 20%로 나눴을 때 남는 고장 건수(대략): 68 건


### 한계 3가지 (실행 결과 기준)

| 한계 | 실행으로 확인한 근거 |
|---|---|
| 고장 건수가 절대적으로 적다 | 전체 고장 339건(3.39%). 시험용 20%로 나누면 약 68건만 남아, 재현율·정밀도가 나눌 때마다 크게 흔들릴 수 있다 |
| 조건 열끼리 정보가 겹친다(다중공선성) | `Air temperature`↔`Process temperature` 상관계수 0.876, `Rotational speed`↔`Torque` 상관계수 -0.875 — 조건 열 5개지만 실질 독립 정보량은 그보다 적다 |
| 지나치게 깨끗한(시뮬레이션형) 데이터 | 결측치 0건, 중복 행 0건 — SECOM(결측치 다수)·콜드체인(155건 결측) 같은 실제형 데이터와 달리 노이즈가 없어, 여기서 잘 나온 성능이 실제 현장 데이터에서도 그대로 나올지는 별개 문제다 |

# 실습 17 - 모델링 계획표

## 모델링 계획

| # | 하는 일 | 내 경우엔 | 막힐 것 같나 |
|---|---|---|---|
| 1 | 파일 불러오기 |16_machine-failure.csv 불러오기  |없음 |
| 2 | 빈칸 채우기 |결측치 0건이라 채울 게 없음  |없음 |
| 3 | 결과 열을 숫자로 |Machine failure를 정상=0, 고장=1로  |없음 |
| 4 | 입력과 정답 나누기 |X는 숫자 측정 열 5개(Air temperature [K], Process temperature [K], Rotational speed [rpm], Torque [Nm], Tool wear [min]), y는 Machine failure  |Air/Process 온도, 회전수/토크가 서로 겹치는 정보라 나중에 헷갈릴 수 있음 |
| 5 | 학습용·시험용 나누기 | [8대 2, 불량 비율을 양쪽에 맞춰서] |고장이 339건뿐이라 시험용엔 약 68건만 남음 |
| 6 | 기준 모델 점수 재기 | [전부 양품이라 답하는 모델] |기준 정확도가 96.6%로 이미 높아서 비교가 어려울 수 있음 |
| 7 | 모델 하나 학습시키기 | [표준화한 뒤 로지스틱 회귀] |Air/Process 온도, 회전수/토크 다중공선성 때문에 계수가 불안정할 수 있음 |
| 8 | 결과 읽기 | [기준 모델과 나란히 놓고 비교] |고장 건수가 적어 재현율·정밀도가 들쭉날쭉해 보일 수 있음  |

### 볼 숫자와 이유

| 언제 | 볼 숫자 | 이유 |
|---|---|---|
| 5. 나눈 직후 | 학습용/시험용 각각의 건수와 그 안의 고장 건수 | 8:2로 나눴을 때 양쪽에 고장 비율(3.39%)이 비슷하게 들어갔는지 확인 — 층화추출이 제대로 됐는지 보는 것 |
| 6. 기준 모델 | 기준 모델 정확도 (전부 정상이라고만 답했을 때, 약 96.6%) | 이게 "아무것도 안 배운" 최저 기준선. 나중에 만든 모델이 이 숫자보다 못하면 의미가 없음 |
| 7~8. 모델 결과 | 혼동행렬 4칸 — 맞힌 정상 / 헛경보 / 놓친 고장 / 잡은 고장 | 정확도만 보면 96%대로 기준 모델과 비슷해 보여도, 실제로 고장을 몇 건 잡았는지는 이 4칸을 봐야 알 수 있음 |
| 7~8. 모델 결과 | 재현율 (전체 고장 중 몇 %를 잡았나) | 이 프로젝트에서 가장 중요한 숫자 — 고장을 놓치면 설비가 실제로 멈추는 피해로 이어짐 |
| 7~8. 모델 결과 | 정밀도 (고장이라 한 것 중 몇 %가 진짜였나) | 헛경보가 너무 많으면 정비 담당자가 매번 확인하러 다녀야 해서 예측을 안 믿게 됨 |
| 7~8. 모델 결과 | F1 (재현율과 정밀도를 함께 본 값) | 재현율만 올리면 정밀도가 떨어지는 trade-off가 있어서, 둘을 한 숫자로 같이 볼 때 씀 |

### 데이터 누수(leakage) 점검 — 측정 직후 시점 기준

숫자로 된 열: `Air temperature [K]`, `Process temperature [K]`, `Rotational speed [rpm]`, `Torque [Nm]`, `Tool wear [min]` (5개)

| 기준 | 해당하는 열 |
|---|---|
| 1. 일련번호/순번처럼 뜻 없는 숫자 | 없음 |
| 2. 판단 시점(측정 직후)엔 아직 안 찍혀 있을 값 | 없음 |
| 3. `Machine failure`를 다른 형태로 바꿔놓은 것 | 없음 |

**왜 그렇게 봤는지**
- **1번 아님**: `Rotational speed [rpm]` 예시값이 1551, 1408, 1498, 1433...으로 들쭉날쭉하고 10000행 중 유일값이 941개. `Tool wear [min]`도 0, 3, 5, 7, 9...로 실제 공구 사용시간처럼 보임 — 순번이라면 0,1,2,3처럼 규칙적으로 늘어나야 하는데 그런 패턴이 없다.
- **2번 아님**: 5개 열 모두 "자동 측정"이 재는 그 값(온도·회전수·토크·공구마모)이라, "값이 표에 올라온 직후"라는 판단 시점엔 이미 다 찍혀 있는 값이다.
- **3번 아님**: 결과 열 `Machine failure`는 정상/고장 문자열이고, 5개 숫자 열은 서로 다른 물리량(온도·회전수·토크·마모)이다. 이름·값 어디에도 결과 열을 다시 적어놓은 흔적이 없다.

# 실습 18 - 모델링 v1

## 모델링 v1

### 문법 노트 - 코드에서 찾을 이름

| 이름 | 뜻 |
|---|---|
| features | 모델에 넣는 입력 열 목록 |
| label | 정답표 - 찾으려는 쪽이면 1, 아니면 0 |
| X / y | 입력 / 정답 |
| X_train, X_test | 학습용 입력 / 시험용 입력 |
| y_train, y_test | 학습용 정답 / 시험용 정답 |
| baseline | 기준 모델 - 학습 없이 전부 아니라고 답하는 것 |
| model | 학습시킨 모델 |
| y_pred | 모델이 내놓은 답 |
| precision / recall | 정밀도 / 재현율 |

### 내 갈래

- 갈래 : [A - 맞히기(결과 열이 있다)]
- 결과 열 : [Machine failure] / 관심 갈래 : [고장]
- 이 기준으로 정한 이유 : [현장에서 걸러내려는 게 고장 셀이라서]

## Step 4. 모델링 v1

In [18]:
# 1. 파일을 불러온다
df = pd.read_csv("../../data/16_machine-failure.csv")

# 2. 숫자로 된 열만 입력으로 쓴다. 결과 열(Machine failure)은 답이므로 뺀다
num_cols = df.select_dtypes(include="number").columns.tolist()
X = df[num_cols].copy()

# 3. 입력 열의 빈칸은 그 열의 중앙값으로 채운다
X = X.fillna(X.median())

# 4. 결과 열이 "고장"이면 1, 아니면 0인 label 열을 만든다
df["label"] = (df["Machine failure"] == "고장").astype(int)
y = df["label"]

print("입력 열:", num_cols)
print("입력 크기:", X.shape)
print("label 분포:", y.value_counts().to_dict())

입력 열: ['Air temperature [K]', 'Process temperature [K]', 'Rotational speed [rpm]', 'Torque [Nm]', 'Tool wear [min]']
입력 크기: (10000, 5)
label 분포: {0: 9661, 1: 339}


In [19]:
import numpy as np
from sklearn.model_selection import train_test_split

# 5. 학습용·시험용을 8:2로 나눈다. stratify=y로 양쪽의 1 비율을 같게 맞춘다
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, stratify=y, random_state=42
)

print(f"학습용 {len(X_train)}건 (고장 {y_train.sum()}건, {round(y_train.mean()*100,2)}%)")
print(f"시험용 {len(X_test)}건 (고장 {y_test.sum()}건, {round(y_test.mean()*100,2)}%)")

# 6. 기준 모델 점수 - 학습 없이 전부 0(정상)이라고만 답했을 때의 정확도
기준예측 = np.zeros(len(y_test), dtype=int)
기준정확도 = (기준예측 == y_test).mean()
print("기준 모델 정확도:", round(기준정확도 * 100, 2), "%")

학습용 8000건 (고장 271건, 3.39%)
시험용 2000건 (고장 68건, 3.4%)
기준 모델 정확도: 96.6 %


In [20]:
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, confusion_matrix, precision_score, recall_score

# 7. 표준화를 한 뒤 로지스틱 회귀 모델을 학습시킨다 (학습용으로만 fit)
scaler = StandardScaler()
X_train_스케일 = scaler.fit_transform(X_train)
X_test_스케일 = scaler.transform(X_test)  # 시험용은 학습에 쓰지 않고 변환만 한다

model = LogisticRegression()
model.fit(X_train_스케일, y_train)
예측 = model.predict(X_test_스케일)

# 8. 시험용으로 일곱 줄을 출력한다
내모델정확도 = accuracy_score(y_test, 예측)
맞힌정상, 헛경보, 놓친고장, 잡은고장 = confusion_matrix(y_test, 예측).ravel()
지목한건수 = (예측 == 1).sum()
정밀도 = precision_score(y_test, 예측, zero_division=0)
재현율 = recall_score(y_test, 예측, zero_division=0)

print("기준 모델 정확도:", round(기준정확도 * 100, 2), "%")
print("내 모델 정확도:", round(내모델정확도 * 100, 2), "%")
print("지목한 건수:", 지목한건수)
print("그중 진짜:", 잡은고장)
print("시험용의 실제 고장 건수:", y_test.sum())
print("정밀도:", round(정밀도, 3))
print("재현율:", round(재현율, 3))

기준 모델 정확도: 96.6 %
내 모델 정확도: 96.85 %
지목한 건수: 13
그중 진짜: 9
시험용의 실제 고장 건수: 68
정밀도: 0.692
재현율: 0.132


### 알아낸 것 (실행 결과 기준)

| 항목 | 값 |
|---|---|
| 기준 모델 정확도 | 96.6% |
| 내 모델 정확도 | 96.85% |
| 지목한 건수 | 13 |
| 그중 진짜 | 9 |
| 시험용의 실제 고장 건수 | 68 |
| 정밀도 | 0.692 |
| 재현율 | 0.132 |

기준 모델(전부 정상)보다 정확도는 0.25%p밖에 안 올랐고, 68건의 실제 고장 중 9건만 잡아서(재현율 0.132) 놓친 게 많다 — 앞서 "한계"에서 짚었던 대로 시험용에 고장이 68건뿐이고, 조건 열끼리 겹치는 정보가 많은 게 여기서 그대로 드러난 걸로 보인다.

In [21]:
# 앞 모델(model)은 그대로 두고, class_weight="balanced"만 다르게 준 모델을 새로 만든다
model_가중치 = LogisticRegression(class_weight="balanced")
model_가중치.fit(X_train_스케일, y_train)
예측_가중치 = model_가중치.predict(X_test_스케일)

지목건수_가중치 = (예측_가중치 == 1).sum()
그중진짜_가중치 = ((예측_가중치 == 1) & (y_test == 1)).sum()
재현율_가중치 = recall_score(y_test, 예측_가중치, zero_division=0)

print(f"{'모델':10} {'지목 건수':>8} {'그중 진짜':>8} {'재현율':>8}")
print(f"{'손 안 댐':10} {지목한건수:>8} {잡은고장:>8} {round(재현율,3):>8}")
print(f"{'가중치':10} {지목건수_가중치:>8} {그중진짜_가중치:>8} {round(재현율_가중치,3):>8}")

모델            지목 건수    그중 진짜      재현율
손 안 댐            13        9    0.132
가중치             404       57    0.838


### 결과 정리 (실행 결과 기준)

| 모델 | 지목 건수 | 그중 진짜 | 재현율 |
|---|---|---|---|
| 손 안 댐 | 13 | 9 | 0.132 |
| 가중치 | 404 | 57 | 0.838 |

가중치를 주니 재현율이 0.132→0.838로 크게 오르지만, 지목 건수도 13건→404건으로 급증합니다(그만큼 헛경보도 늘었다는 뜻).

### 비전문가용 설명

이 모델은 2000대의 시험용 설비 중에서 "고장 날 것 같은 설비"를 골라내는 일을 합니다.

- **손 안 댄 모델**은 신중한 편입니다. 13대만 "위험하다"고 지목했는데, 그중 9대가 진짜 고장이었습니다. 다만 실제 고장은 68대였으니, 나머지 59대는 놓쳤다는 뜻입니다(재현율 0.132 = 13.2%만 잡았다).
- **가중치를 준 모델**은 훨씬 더 많이 의심합니다. 404대를 "위험하다"고 지목했고, 그중 57대가 진짜 고장이었습니다. 68대 중 57대를 잡았으니 훨씬 많이 잡은 셈입니다(재현율 0.838 = 83.8%).

그런데 여기엔 대가가 있습니다. 404대 중 57대만 진짜고 나머지 347대(404-57)는 멀쩡한데 "위험하다"고 잘못 지목한 것(헛경보)입니다. 손 안 댄 모델의 헛경보는 13대 중 4대(13-9)뿐이었으니, 헛경보가 4건에서 347건으로 크게 늘어난 셈입니다.

**한 줄로 정리하면**: 가중치를 주면 "진짜 고장을 놓치는 일"은 훨씬 줄어들지만, 대신 "멀쩡한 설비를 괜히 의심하는 일"이 훨씬 늘어납니다. 그래서 어느 쪽을 쓸지는 "고장을 놓쳤을 때의 손해"와 "괜히 점검하러 갔을 때의 수고" 중 어느 쪽이 회사 입장에서 더 부담스러운지에 달려 있습니다.

# 실습 19 - 결과 한 장

## v1 결과

| | 정확도 | 지목 | 그중 진짜 | 정밀도 | 재현율 |
|---|---|---|---|---|---|
| 기준 모델 (전부 합격) | 96.6% | 0건 | 0건 | 0.000 | 0.000 |
| 내 모델 v1 | 96.85% | 13건 | 9건 | 0.692 | 0.132 |


- 시험용에 실제로 있던 불합격 : **68건**
- 한 줄 요약 : 정확도는 기준보다 **0.25**%p 올랐고 지목한 13건 중 9건이 실제 불합격이었다(정밀도 0.692). 다만 실제 **68**건 중 **59**건은 못 잡았다.(재현율 0.132<--- 9건(진짜)/68건(시험용 실제 불합격))

### 다음에 할 일

1. [드문 쪽을 무겁게 세는 설정으로 지목 건수를 늘려보기]
2. [지목을 늘렸을 때 헛걸음이 얼마나 느는지 같이 재기]
3. [재검사 한 건에 드는 비용을 알아보고, 어느 쪽이 나은지 정하기]

### 카드에 옮겨 적을 것

**3번 칸 - 사용 데이터**
[셀 검사 기록 / 10000행 x 7열 / 결과 열은 정상·고장 / 고장 339건 (3.39%)]

**5번 칸 - 결과 수치**
[전부 합격이라 답하는 기준 모델 96.6% 대비 96.85%.
지목한 13건 중 9건이 실제 고장이었으나(정밀도 0.692),
실제 68건 중 59건은 못 잡았다(재현율 0.132). 그래서 정확도만으로는 판단할 수 없는 문제였다]

### v1 결과 · 카드 내용 다듬은 버전 (숫자는 그대로, 어려운 말만 풀어씀)

**v1 결과**

| | 정확도(전체 중 맞힌 비율) | 지목 | 그중 진짜 | 정밀도(지목한 것 중 진짜였던 비율) | 재현율(실제 있던 것 중 잡아낸 비율) |
|---|---|---|---|---|---|
| 기준 모델 (전부 합격) | 96.6% | 0건 | 0건 | 0.000 | 0.000 |
| 내 모델 v1 | 96.85% | 13건 | 9건 | 0.692 | 0.132 |

- 시험용에 실제로 있던 불합격 : **68건**
- 한 줄 요약 : 정확도(전체 중 맞힌 비율)는 기준보다 **0.25**%p 올랐고, 지목한 13건 중 9건이 실제 불합격이었다(정밀도, 즉 지목한 것 중 진짜였던 비율 0.692). 다만 실제 **68**건 중 **59**건은 못 잡았다(재현율, 즉 실제 있던 것 중 잡아낸 비율 0.132 = 9건 ÷ 68건).

**카드에 옮겨 적을 것**

3번 칸 - 사용 데이터
셀 검사 기록 / 10000행 x 7열 / 결과 열은 정상·고장 / 고장 339건 (3.39%)

5번 칸 - 결과 수치
전부 합격이라 답하는 기준 모델(정확도 96.6%) 대비 정확도 96.85%. 지목한 13건 중 9건이 실제 고장이었으나(정밀도, 즉 지목한 것 중 진짜였던 비율 0.692), 실제 68건 중 59건은 못 잡았다(재현율, 즉 실제 있던 것 중 잡아낸 비율 0.132). 그래서 정확도(전체 중 맞힌 비율)만으로는 판단할 수 없는 문제였다.

- 오늘 정한 문제 : [셀 검사 측정값으로 정밀 재검사 대상을 먼저 고르는 것]
- 쓴 데이터 : [셀 검사 기록 10000행 x 7열 / 불합격 339건 3.39%]
- 오늘 나온 숫자 : [기준 96.6% / v1 96.85% / 지목 13건 중 진짜 9건 / 재현율 0.132]
- 다음에 손댈 자리 : [지목 건수를 늘리는 설정]

### 도전 - 그래서 무엇을 하자는 것인가

- [모델이 고장 가능성이 높다고 지목한 셀은 최종 검사로 넘기기 전에 정밀 재검사로 돌리자는 것.
  지금 v1로는 실제 68건 중 9건만 걸러지므로, 이 상태로 현장에 걸기에는 이르다]